lidlplus/html_receipt.py

```diff
@@ -0,0 +1,65 @@
+from typing import Any
+import re
+
+import lxml.html as html
+
+
+VAT_TYPE_LINE_ENDING_PATTERN = re.compile(r" [A-Z]$")
+
+
+def parse_html_receipt(date: str, html_receipt: str) -> dict[str, Any]:
+    dom = html.document_fromstring(html_receipt)
+
+    receipt = {
+        "date": date,
+        "itemsLine": [],
+    }
+    for node in dom.xpath(r".//span[starts-with(@id, 'purchase_list_line_')]"):
+        if "class" not in node.attrib:
+            if not VAT_TYPE_LINE_ENDING_PATTERN.search(node.text):
+                continue
+
+            *name_parts, price = node.text[:-2].split()
+            receipt["itemsLine"].append(
+                {
+                    "name": " ".join(name_parts),
+                    "originalAmount": price,
+                    "isWeight": True,
+                    "discounts": [],
+                }
+            )
+        elif node.attrib["class"] == "currency":
+            receipt["currency"] = {"code": node.text.strip(), "symbol": node.attrib["data-currency"]}
+        elif node.attrib["class"] == "article":
+            if node.text.startswith(" "):
+                continue
+
+            quantity_text = node.get("data-art-quantity")
+            if quantity_text is None:
+                is_weight = False
+                quantity = 1
+            elif "," in quantity_text:
+                is_weight = True
+                quantity = quantity_text
+            else:
+                is_weight = False
+                quantity = quantity_text
+
+            receipt["itemsLine"].append(
+                {
+                    "name": node.attrib["data-art-description"],
+                    "currentUnitPrice": node.attrib["data-unit-price"],
+                    "isWeight": is_weight,
+                    "quantity": quantity,
+                    "discounts": [],
+                }
+            )
+        elif node.attrib["class"] == "discount":
+            discount = abs(parse_float(node.text.split()[-1]))
```

Comment:
This throws an IndexError when I'm running it because some of the span elements contain just white text so the `node.text.split()` returns an empty list. 

Here's the HTML on my receipt:

```<span id="purchase_list_line_3" class="discount" data-promotion-id="promo_id">  </span>
<span id="purchase_list_line_3" class="discount css_bold" data-promotion-id="promo_id">   Coupon Plus reward</span>
<span id="purchase_list_line_3" class="discount" data-promotion-id="promo_id">      </span>
<span id="purchase_list_line_3" class="discount" data-promotion-id="promo_id">     </span>
<span id="purchase_list_line_3" class="discount css_bold" data-promotion-id="promo_id">-0.69</span>```

Reply:
Man, they just can't be consistent... Thank you for providing another data point with which we can figure out all the formats they use for this! I'll try to implement the format you provided once I have time to actually do this though... You can always suggest changes and I'll be happy to use them of course.

For now I'm thinking a regex searching for something like `-\d+[\.,]\d{2}$` would be best.

Btw shouldn't the code currently fail in parsing the first line's `reward` word as float instead of the whitespace-split-index-error that you're describing?

Reply:
No rush to implement this. I would have done it myself but I was hesitant to suggest changes cause I'm still trying to understand what's happening. I will for sure once I get more familiar with the project. :)

I think because the class of the first line is `discount ccs_bold` instead of just `discount `it's not parsed at all. 

So, does the HTML differ from country to country? Or does it depend on the coupon you use and whether it's a percentage/flat discount? That's the only receipt with a coupon I have so that's my only data point :/

Reply:
> I was hesitant to suggest changes cause I'm still trying to understand what's happening. I will for sure once I get more familiar with the project.

No worries :) This is not even that tied to this specific project, but more to the actual lidl API since AFAIK there's no public documentation for it and people just somehow reverse engineered it.

> I think because the class of the first line is discount ccs_bold instead of just discount it's not parsed at all.

Right, of course, missed that.

> So, does the HTML differ from country to country? Or does it depend on the coupon you use and whether it's a percentage/flat discount?

There's definitely *some* difference for *some* reason. See [diogotcorreia/lidl-to-grocy's lidl/test/receipt.html](https://github.com/diogotcorreia/lidl-to-grocy/blob/be2ae2b444fe10c15a048f38585e1717cdd3cadd/lidl/test/receipt.html). It uses I think the same format as what I saw and implemented in this PR. It's weird that your receipt is different, but we'll probably have to implement a common parsing for all possible formats. Currently I'm blocked on #23 though so I can't verify if anything changed recently in my receipts, but in my lidl-plus android app I don't see any discounts as bold as you probably would.

---
We could probably do something like this to support both formats:
```py
if ...:
    ...
elif {"discount", "css_bold"}.issubset(node.attrib["class"].split()) and try_parse_float(node.text):
    ...
elif node.attrib["class"] == "discount":
    ...
```